distributed/shuffle/_shuffle.py

```diff
@@ -489,10 +489,14 @@ async def _receive(self, data: list[tuple[int, bytes]]) -> None:
             self._exception = e
             raise
 
+    @property
+    def npartitions(self) -> int:
```

Comment:
A few comments:
* Is there a reason not to calculate this once at initialization?
* Personally, I'd find `max(self.worker_for)` an easier calculation and we're interested in the maximum value, not the number of partitions (I _think_ there could be gaps in the partition numbers if we selected data beforehand, but I'm not 100% sure right now).
* `npartitions` is typically used for the number of output partitions in this context. `max_partition_input` would match what we're looking for here.

Reply:
> Is there a reason not to calculate this once at initialization?

laziness. This was closer to the code I was modifying than `__init__` ;)

> Personally, I'd find max(self.worker_for) an easier calculation and we're interested in the maximum value, not the number of partitions (I think there could be gaps in the partition numbers if we selected data beforehand, but I'm not 100% sure right now).

I wasn't sure any more what the keys in worker_for were doing. I'll rechenk. Regarding the gaps in partition numbers, I believe this is impossible since this is a `range` over `npartitions_input` https://github.com/fjetter/distributed/blob/4bea59f704f5f2a566a176fb7d72a71d1ca9c1e1/distributed/shuffle/_shuffle.py#L276
if max works, I'll take that and remove the property again

> npartitions is typically used for the number of output partitions in this context. max_partition_input would match what we're looking for here.

I think I'll just remove the property again (was reading too much dask-expr code recently )

Reply:
> I wasn't sure any more what the keys in worker_for were doing. I'll rechenk. Regarding the gaps in partition numbers, I believe this is impossible since this is a `range` over `npartitions_input`

Ah, you're right! We could also go with `npartitions_input` then which matches the `P2PShuffleLayer`.

AFAIR, `worker_for` is the inverse mapping of `partitions_of`, i.e., it maps partition IDs (key) to designated workers (value).

Reply:
> AFAIR, worker_for is the inverse mapping of partitions_of, i.e., it maps partition IDs (key) to designated workers (value).

...which already reveals my flawed thinking. Neither of the two should work, both are mappings for the _output_ partitions. We want input partitions though. So I guess we will have to pipe the `npartitions_input` from the layer to the transfer tasks.

Reply:
I went with `max`. This trades a little compute time with a little better decoupling from that "partition ID are consecutive integers without gaps" assumption

Reply:
Ok I went the other way. Every np.full uses the smallest width possible for the given partition. I will trust arrow to take care of upcasting where necessary.

I extended one of the tests to test partition ranges beyond unit8 range to make sure this edge case is hit.